File: Arrays/3719_longest_balanced_subarry_i/solution.py
class Solution:
    def longestBalanced(self, nums: List[int]) -> int:
        max_length = 0
        n = len(nums)

        for start in range(n):
            distinct_evens = set()
            distinct_odds = set()

            for end in range(start, n):
                value = nums[end]
                if value % 2 == 0:
                    distinct_evens.add(value)
                else:
                    distinct_odds.add(value)

                if len(distinct_evens) == len(distinct_odds):
                    max_length = max(max_length, end - start + 1)

        return max_length

File: Arrays/3719_longest_balanced_subarry_i/Solution.java
class Solution {
    public int longestBalanced(int[] nums) {
        int maxLength = 0;
        int n = nums.length;

        for (int start = 0; start < n; start++) {
            Set<Integer> distinctEvens = new HashSet<>();
            Set<Integer> distinctOdds = new HashSet<>();

            for (int end = start; end < n; end++) {
                int value = nums[end];
                if (value % 2 == 0) {
                    distinctEvens.add(value);
                } else {
                    distinctOdds.add(value);
                }

                if (distinctEvens.size() == distinctOdds.size()) {
                    maxLength = Math.max(maxLength, end - start + 1);
                }
            }
        }
        return maxLength;
    }
}
